cpu/stm32_common/periph/rtc.c

```diff
@@ -322,4 +326,231 @@ void ISR_NAME(void)
+
+    /* oscillator clock used as RTC clock */
+    RCC->BDCR |= RCC_BDCR_RTCEN | RCC_BDCR_RTCSEL_LSE;
+
+    /* turn on LSE crystal */
+    RCC->BDCR |= RCC_BDCR_LSEON;
+    while ((RCC->BDCR & RCC_BDCR_LSEON) != RCC_BDCR_LSEON) {}
+
+    /* calibration clock from 0 to 0x7F */
+    BKP->RTCCR |= 0;
+
+    /* second interrupt is disabled. */
+    RTC->CRH &= ~RTC_CRH_SECIE;
+
+    _rtc_enter_config_mode();
+
+    /* if the input clock frequency (fRTCCLK) is 32.768 kHz, write 7FFFh in this register to get a signal period of 1 second. */
+    RTC->PRLH = 0;
+    RTC->PRLL = 0x7FFF;
+
+    _rtc_exit_config_mode();
+
+    /* wait registers synchronize flag */
+    RTC->CRL &= (uint16_t)~RTC_CRL_RSF;
+    while((RTC->CRL & RTC_CRL_RSF) != RTC_CRL_RSF) {}
+
+    /* disable backup domain write protection */
+    PWR->CR &= ~PWR_CR_DBP;
+}
+
+static time_t _rtc_get_time(void)
+{
+    return (RTC->CNTH << 16) | RTC->CNTL;
+}
+
+static void _rtc_set_time(time_t counter_val)
+{
+    _rtc_enter_config_mode();
+    RTC->CNTH = (counter_val & 0xffff0000) >> 16;
+    RTC->CNTL = counter_val & 0x0000ffff;
+    _rtc_exit_config_mode();
+}
+
+void rtc_init(void)
+{
+    /* save current time if RTC already works */
+    bool is_rtc_enable = _is_rtc_enable();
+    time_t cur_time = 0;
+    if (is_rtc_enable) {
+        cur_time = _rtc_get_time();
+    }
+
+    /* config RTC */
+    _rtc_config();
+
+    /* restore current time after config */
+    if (is_rtc_enable) {
+        _rtc_set_time(cur_time);
+    }
+}
+
+int rtc_set_time(struct tm *time)
+{
+    time_t timestamp = mktime(time);
+    _rtc_set_time(timestamp);
+
+    DEBUG("%s timestamp=%lu\n", __func__, (uint32_t)timestamp);
+
+    return 0;
+}
+
+int rtc_get_time(struct tm *time)
+{
+    time_t timestamp = _rtc_get_time();
+    localtime_r(&timestamp, time);
+
+    DEBUG("%s timestamp=%lu\n", __func__, (uint32_t)timestamp);
+
+    return 0;
+}
+
+static void _rtc_enable_alarm(void)
+{
+    /* clear alarm flag */
+    RTC->CRL &= ~RTC_CRL_ALRF;
+
+    _rtc_enter_config_mode();
+    RTC->CRH |= (RTC_CRH_ALRIE);
+    _rtc_exit_config_mode();
+}
+
+static void _rtc_disable_alarm(void)
+{
+    _rtc_enter_config_mode();
+    RTC->CRH &= ~(RTC_CRH_ALRIE | RTC_CRH_SECIE);
+    _rtc_exit_config_mode();
+}
+
+static time_t _rtc_get_alarm_time(void)
+{
+    return (RTC->ALRH << 16) | RTC->ALRL;
+}
+
+static void _rtc_set_alarm_time(time_t alarm_time)
+{
+    _rtc_enter_config_mode();
+    RTC->ALRL = (alarm_time & 0x0000ffff);
+    RTC->ALRH = (alarm_time & 0xffff0000) >> 16;
+    _rtc_exit_config_mode();
+}
+
+int rtc_set_alarm(struct tm *time, rtc_alarm_cb_t cb, void *arg)
+{
+    /* enable global RTC interrupt */
+    if (cb && !NVIC_GetEnableIRQ(RTC_IRQn)) {
+        NVIC_EnableIRQ(RTC_IRQn);
+        DEBUG("%s enable RTC IRQ\n", __func__);
+    }
+
+    /* save callback and argument */
+    isr_ctx.cb = cb;
+    isr_ctx.arg = arg;
+
+    /* set wakeup time */
+    time_t timestamp = mktime(time);
```

Comment:
`mktime()` returns `(time_t) -1` on error. In that case `rtc_set_alarm()` should return `-2` according to the API doc.

Btw: I'm personally would favor to always check the arguments for being plausible (e.g. `time` and `cb` should not be `NULL`, `time` should be in the future).

Reply:
I think that `cd` may be `NULL`. When `rtc_set_alarm()` use for wake up after standby callback not need.
Alarm time save in standby mode. In alarm time in past may encode several bytes data. It is possible to save several bytes in the standby mode. Maybe it can be useful for processor who do not have EEPROM.

Reply:
> I think that `cd` may be NULL. When `rtc_set_alarm()` use for wake up after standby callback not need.

You're right. Also, `void isr_rtc(void)` does properly check if `cb` is `NULL` prior to calling, so there is no risk of calling a `NULL`-function pointer :-)

> Alarm time save in standby mode. In alarm time in past may encode several bytes data. It is possible to save several bytes in the standby mode. Maybe it can be useful for processor who do not have EEPROM.

Someone re-purposing an RTC alarm to store and retrieve data would very likely not use the RTC API. But no other RTC implementation in RIOT does check for alarm time in the past nor is it required by the API doc. Therefore it would indeed be more consistent to not check here is well.

Reply:
> In alarm time in past may encode several bytes data. It is possible to save several bytes in the standby mode. Maybe it can be useful for processor who do not have EEPROM.

AFAIK the stm32 have backup memory which is part of the RTC peripheral. Still this shouldn't be accessed through th ertc API, as @maribu stated.

Reply:
I had some code lying around for using the backup registers, please see #11369.

Reply:
> AFAIK the stm32 have backup memory which is part of the RTC peripheral. Still this shouldn't be accessed through th ertc API, as @maribu stated.

STM32f103 do not have EEPROM and RTC->BKP0R. Use `rtc_set_alarm()` is not good hack, but this may be one of the last chances.

Reply:
> STM32f103 do not have EEPROM and RTC->BKP0R. Use rtc_set_alarm() is not good hack, but this may be one of the last chances.

There is nothing wrong with resorting to `RTC->BKP0R` for storing non-RTC data. But I strongly believe that the RTC-API is not to be used for that. Instead a different API could be provided to use that, if someone wants to do that. (Feel free to open a separate PR to just do that, if you have a use case for that. This PR should also handle that the register is not used simultaneously as RTC alarm and as storage - maybe as simple as having `rtc_set_alarm()` return `-1` if the RTC-storage-API is used.)

Reply:
IMHO that would best be integrated into https://github.com/RIOT-OS/RIOT/pull/11369 to have a consistent API. Maybe a pseudo module could be used to unlock using the RTC register for just that.